### ISSUE TITLE
ARPACK randomly not converging during hwe_normalized_pca

### ISSUE THREAD

**Issue by iris-garden** [iris-garden/test-process#2725]
<h2>olszewskip said:</h2>
<p>Hi!<br>
I’m getting</p>

```python
Error summary: IllegalStateException: ARPACK returns non-zero info = 1 Maximum number of iterations taken. (Refer ARPACK user guide for details)

```

<p>after calling <code>hwe_normalized_pca(mt.GT, k=2)</code> and I have trouble pinpointing the cause of the problem. What’s worse, when I’m calling this function several times in a row on the same input, sometimes it errors out and sometimes it returns without an error. Is that even possible with a proper installation? I have now made myself a jupyter notebook with the following single cell just to prove to myself that I’m not going mad:</p>

```python
hl.init(quiet=True)
hl.set_global_seed(42)
mt = hl.read_matrix_table(my_path)
for i in range(12):
    try:
        hl.hwe_normalized_pca(
            mt.GT,
            k = 2
        )
        print(f"{i} success")
    except:
        print(f"{i} error")

```

<p>and it prints</p>

```python
0 success
1 success
2 success
3 success
4 error
5 success
6 success
7 success
8 success
9 error
10 success
11 success

```

<p>where <code>mt</code> is a product of a several filtering steps performed on a genuine vcf.<br>
Also, when I do</p>

```python
hl.pca(
    hl.int(hl.is_defined(mt.GT)),
    k = 2
)

```

<p>on my <code>mt</code>, I always immediately get another kind of error:<br>
<code>Error summary: HailException: Cannot create RowMatrix: filtered entry at row 8111 and col 9</code>,<br>
but I don’t know if it is related to the problem with <code>hwe_normalizd_pca</code>.<br>
Any suggestions would be appreciated:</p>
<ol>
<li>What qualities of any <code>mt</code> could in general cause the <code>hwe_normalized_pca</code> to not converge?</li>
<li>Should I suspect that there’s something wrong with my installation? I’m running hail locally, on a laptop.</li>
<li>Shape of <code>mt</code> is (18205, 39), its only non-key field is GT, and the col-keys are generic strings, so I’d be happy to share it.</li>
</ol>

<h2>danking said:</h2>
<p>You could make normal PCA work with <code>hl.or_else</code> to set a value for missing entries.</p>
<p>There’s almost certainly no issue with your installation. I can only speculate as to the issues, but I’d guess that your matrix is exhibiting some kind of numerical instability that makes ARPACK’s SVD routines occasionally not converge. Your dataset is quite small. You might try collecting it locally and converting it to a numpy array for use with sklearn.decomposition.PCA. If that succeeds, it’s possible Hail could use whatever parameters sklearn is using to call ARPACK and achieve more consistent results.</p>

<h2>olszewskip said:</h2>
<p>Thanks, danking, for the response!<br>
It’s helpful to hear that reinstalling probably won’t help.<br>
Yup, the dataset is small, partly because I was trying to take a subset of it which still throws the same error as the original <img src="https://emoji.discourse-cdn.com/twitter/slight_smile.png?v=9" title=":slight_smile:" class="emoji" alt=":slight_smile:"><br>
Not sure where I would put <code>hl.or_else</code> in <code>hl.int(hl.is_defined(mt.GT))</code>. But, swapping <code>mt.GT</code> for <code>mt.unfilter_entries().GT</code> made the <code>hl.pca</code> return without an error. Similarly</p>

```python
from sklearn.decomposition import PCA
mt_2 = mt.unfilter_entries()
df = mt_2.select_entries(
    n = hl.int(hl.is_defined(mt_2.GT))
).key_rows_by().select_rows().make_table().to_pandas()
PCA(2).fit_transform(df)

```

<p>returns without an error, but it would complain about NaN without the <code>unfilter_entries</code> (same with say <code>n = hl.or_else(mt_2.GT.n_alt_alleles(), 0)</code>).<br>
Still, <code>unfilter_entries</code> does not help with the <code>hwe_normalized_pca</code>.<br>
Perhaps I’m having a problem with variants that are far from hw equilibrium. E.g. if I filter out low hw p-value before calling <code>hwe_normalized_pca</code> in my jupyter cell:</p>

```python
mt = hl.read_matrix_table("debugging/wierdness.mt")
mt = mt.filter_rows(
    hl.agg.hardy_weinberg_test(mt.GT).p_value > 1e-6
)
...

```

<p>it throws away 25 variants and prints</p>

```python
0 success
1 success
2 success
3 error
4 success
5 success
6 success
7 success
8 success
9 success
10 success
11 success

```

<h2>danking said:</h2>
<p>It sounds to me like hardy-weinberg normalization is producing a matrix that is likely to be poorly conditioned causing ARPACK to not converge.</p>
<p>Let’s compare apples to apples:</p>

```python
import hail as hl
mt = hl.read_mtarix_table('debugging/wierdness.mt')
mt_for_comparison = hl.methods.pca.hwe_normalize(mt.GT)
# ... run mt_for_comparison through sklearn PCA and hl.pca

```

<h2>olszewskip said:</h2>
<p>How would I import and use the <code>hwe_normalize</code> function? It doesn’t seem to be a part of Hails interface. The following fails</p>

```python
from hail.methods.pca import hwe_normalize
mt.select_entries(
    n = hwe_normalize(mt.GT)
)

```

<p>at <code>select_entries</code> with<br>
<code>ExpressionException: Cannot combine expressions from different source objects.</code>.</p>

<h2>johnc1231 said:</h2>
<p>This is not documented because it’s an internal method. This conversation may end up being an impetus for us to document it and expose it. The usage is similar to what Dan did above, but you’d need to import it the way you did. So:</p>

```python
from hail.methods.pca import hwe_normalize

norm_gt = hwe_normalize(mt.GT)

# note it's not a matrix table, its an entry expression. You can export it for sklearn though using:

norm_gt.export('norm_gt.tsv')

```
